### ISSUE TITLE
resolveJupyterServer is invoked even if the server is already fully resolved

### ISSUE THREAD

**Issue by rebornix** [microsoft/vscode-jupyter#14221]
Testing #14201

Every time when we open the kernel picker and go to the server, `resolveJupyterServer` is called again, even if the server is already connected and it's providing kernels fine (so it's not a reconnect).

**Comment by rebornix** [comment]
`resolveJupyterServer` is also invoked multiple times when running cells.

**Comment by DonJayamanne** [comment]
Unfortunately we do not cache the resolved Jupyter server information, (thtas upto the owning extension)
I think its totally fine to call this again.
The fact that we call this again might not be great, 

> resolveJupyterServer is also invoked multiple times when running cells.

Yes this is the existing code, unfortunately this will be there for some time, will go away when we fully migrate the code to the new provider and then remove a lot of the old code

The root cause is the old code would resolve the server just to get the display name (as the display name was part of the server information returned by the provider).

Closing this as there's nothing to be done here and this will go away when we fully migrate to the new API.
TLDR = caused by debt in old API.

**Comment by DonJayamanne** [comment]
@rebornix 
Once again we do not cache the resolved Jupyter server information, (thtas upto the owning extension)

**Comment by rebornix** [comment]
> Unfortunately we do not cache the resolved Jupyter server information, (thtas upto the owning extension)
I think its totally fine to call this again.

I think it has nothing to do with caching. Jupyter extension doesn't need to request server info again if it's already resolved and the server is still valid/accessible. Note that it's not asking Jupyter to cache the server info for scenarios like window reload, it's simply expecting Jupyter to not ask `resolve*` again when the server info is just hot in the same session.

> TLDR = caused by debt in old API.

Reopening as we may want to tackle this as debt.

**Comment by DonJayamanne** [comment]
> Reopening as we may want to tackle this as debt.

It will get tacked, but only when we move to the full API. i'm not too keen on re-factoring a lot of the code to resolve this.
Unless you think its crucial.

> Note that it's not asking Jupyter to cache the server info for scenarios like window reload, it's simply expecting Jupyter to not ask resolve* again when the server info is just hot in the same sessio

But the problem is we do not know when the auth information/token may have expired.
E.g. its possible the 3rd part returns auth information that is only valid for 60s, and after that we need to re-query it again.
In fact we did have such code in the past, we have since removed this as this ws completely unnecessary (& confirmed with AzML as well, perhaps only required at some point in time)

Agreed lets chat and we can tackle in debt (my plan was to address this when we mvoed to the new API completely), at which point we would never make these bogus requests, right now its just a little too messy to fix that.

**Comment by DonJayamanne** [comment]
> it's simply expecting Jupyter to not ask resolve* again when the server info is just hot in the same session.

We might want to re-vist/discuss this.
There are problems with this approach.
Assume the server was resolved , then we wait for 1 hour.
Shoudl we use the same resolve server? Well this means, Jupyter extension needs to cache (store this resolved information in memory)., 
This is why i used the term `cached` in my earlier messages

Cacing/storing in memory is fraught with issues.
What if the server auth has expired, we have no idea what kihnd of auth is used by the servers. AzML used to have a expiry in the past, and Synapse has an expiry of 20 minutes, thus jupyter now needs to keep this in mind as well.

I think its upto the extensions to cache at their end and return the same `fully resolved` servers.
Or we need to expose an API that can allow extensions to tell Jupyter whether the `resolved` information is out dated or not

I personally think the former is simpler

> Jupyter extension doesn't need to request server info again if it's already resolved and the server is still valid/accessible.

Basically it comes down to, how does Jupyter know that it was resolve earlier, without storing this in memory (i.e. caching in a varialbe).
